File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/DiningPhilosophers/DiningPhilosophers_VisualBasic/CircularPanel.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: CircularPanel.xaml.vb
'
'--------------------------------------------------------------------------


Namespace DiningPhilosophers
	Public Class CircularPanel
		Inherits Panel
		Protected Overrides Function MeasureOverride(ByVal availableSize As Size) As Size
			Dim maxChildSize = New Size()
            For Each child As UIElement In InternalChildren
                child.Measure(availableSize)
                If maxChildSize.Width < child.DesiredSize.Width Then
                    maxChildSize.Width = child.DesiredSize.Width
                End If
                If maxChildSize.Height < child.DesiredSize.Height Then
                    maxChildSize.Height = child.DesiredSize.Height
                End If
            Next child
			Return maxChildSize
		End Function

		Protected Overrides Function ArrangeOverride(ByVal finalSize As Size) As Size
			Dim children = InternalChildren.OfType(Of UIElement)().ToArray()
			If children.Length > 0 Then
                Dim midPanel = New Point(finalSize.Width / 2, finalSize.Height / 2)
				Dim maxChild = New Size(children.Max(Function(u) u.DesiredSize.Width), children.Max(Function(u) u.DesiredSize.Height))
				Dim radius = New Size((finalSize.Width - maxChild.Width) / 2, (finalSize.Height - maxChild.Height) / 2)
                Dim arcRadiansPerChild = Math.PI * 2 / children.Length

                Dim curPos = 0
				For Each child In children
					Dim childAngleInRadians = curPos * arcRadiansPerChild
                    Dim childPosition = New Point((Math.Sin(childAngleInRadians) * radius.Width) + (midPanel.X - (child.DesiredSize.Width / 2)),
                                                  (Math.Cos(childAngleInRadians) * radius.Height) + (midPanel.Y - (child.DesiredSize.Height / 2)))
					child.Arrange(New Rect(childPosition, child.DesiredSize))
					curPos += 1
				Next child
			End If
			Return finalSize
		End Function
	End Class
End Namespace

File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/DiningPhilosophers/DiningPhilosophers_CSharp/CircularPanel.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: CircularPanel.xaml.cs
//
//--------------------------------------------------------------------------

using System;
using System.Linq;
using System.Windows;
using System.Windows.Controls;

namespace DiningPhilosophers
{
    public class CircularPanel : Panel
    {
        protected override Size MeasureOverride(Size availableSize)
        {
            var maxChildSize = new Size();
            foreach (UIElement child in InternalChildren)
            {
                child.Measure(availableSize);
                if (maxChildSize.Width < child.DesiredSize.Width) maxChildSize.Width = child.DesiredSize.Width;
                if (maxChildSize.Height < child.DesiredSize.Height) maxChildSize.Height = child.DesiredSize.Height;
            }
            return maxChildSize;
        }

        protected override Size ArrangeOverride(Size finalSize)
        {
            var children = InternalChildren.OfType<UIElement>().ToArray();
            if (children.Length > 0)
            {
                var midPanel = new Point(finalSize.Width / 2, finalSize.Height / 2);
                var maxChild = new Size(children.Max(u => u.DesiredSize.Width), children.Max(u => u.DesiredSize.Height));
                var radius = new Size((finalSize.Width - maxChild.Width) / 2, (finalSize.Height - maxChild.Height) / 2);
                double arcRadiansPerChild = Math.PI * 2 / children.Length;

                int curPos = 0;
                foreach (var child in children)
                {
                    var childAngleInRadians = curPos * arcRadiansPerChild;
                    var childPosition = new Point(
                        (Math.Sin(childAngleInRadians) * radius.Width) + (midPanel.X - (child.DesiredSize.Width / 2)),
                        (Math.Cos(childAngleInRadians) * radius.Height) + (midPanel.Y - (child.DesiredSize.Height / 2)));
                    child.Arrange(new Rect(childPosition, child.DesiredSize));
                    curPos++;
                }
            }
            return finalSize;
        }
    }
}